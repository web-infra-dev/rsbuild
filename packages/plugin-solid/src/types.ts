/**
 * babel-preset-solid options
 *
 * https://github.com/solidjs/solid/blob/main/packages/babel-preset-solid/index.js
 *
 * https://github.com/ryansolid/dom-expressions/blob/main/packages/babel-plugin-jsx-dom-expressions/README.md
 */
export type SolidPresetOptions = {
  /**
   * The name of the runtime module to import the methods from.
   */
  moduleName?: string;
  /**
   * The output mode of the compiler. Can be "dom"(default), "ssr". "dom" is standard output. "ssr" is for server side rendering of strings.
   * @default 'dom'
   */
  generate?: 'dom' | 'ssr';
  /**
   * Indicate whether the output should contain hydratable markers.
   * @default false
   */
  hydratable?: boolean;
  /**
   * Boolean to indicate whether to enable automatic event delegation on camelCase.
   * @default true
   */
  delegateEvents: boolean;
  /**
   * Boolean indicates whether smart conditional detection should be used. This optimizes simple boolean expressions and ternaries in JSX.
   * @default true
   */
  wrapConditionals?: boolean;
  /**
   * Boolean indicates whether to set current render context on Custom Elements and slots. Useful for seemless Context API with Web Components.
   * @default false
   */
  contextToCustomElements?: boolean;
  /**
   * Array of Component exports from module, that aren't included by default with the library. This plugin will automatically import them if it comes across them in the JSX.
   * @default string[]
   */
  builtIns?: string[];
  /**
   * This plugin leverages a heuristic for reactive wrapping and lazy evaluation of JSX expressions.
   * @default 'effect'
   */
  effectWrapper?: string;
  /**
   * Comment decorator string indicates the static expression, used to tell the compiler not to wrap them by effect function
   * @default '@once'
   */
  staticMarker?: string;
  /**
   * Memos let you efficiently use a derived value in many reactive computations. This option indicates the memo function name
   * @default 'memo'
   */
  memoWrapper?: string;
  /**
   * Checks for properly formed HTML by checking for elements that would not be allowed in certain parent elements.
   * @default true
   */
  validate?: boolean;
  /**
   * Removes unnecessary closing tags from the template output.
   * @default true
   */
  omitNestedClosingTags?: boolean;
};
