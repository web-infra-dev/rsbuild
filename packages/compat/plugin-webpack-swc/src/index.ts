export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
