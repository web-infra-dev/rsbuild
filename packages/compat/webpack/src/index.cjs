/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
