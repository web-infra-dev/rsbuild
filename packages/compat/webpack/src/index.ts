export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
