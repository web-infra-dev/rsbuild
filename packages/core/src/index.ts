/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
