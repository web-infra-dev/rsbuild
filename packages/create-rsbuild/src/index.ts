#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import {
  cancel,
  isCancel,
  multiselect,
  note,
  outro,
  select,
  text,
} from '@clack/prompts';
import deepmerge from 'deepmerge';
import { logger } from 'rslog';

function cancelAndExit() {
  cancel('Operation cancelled.');
  process.exit(0);
}

function checkCancel<T>(value: unknown) {
  if (isCancel(value)) {
    cancelAndExit();
  }
  return value as T;
}

function formatTargetDir(targetDir: string) {
  return targetDir.trim().replace(/\/+$/g, '');
}

function pkgFromUserAgent(userAgent: string | undefined) {
  if (!userAgent) return undefined;
  const pkgSpec = userAgent.split(' ')[0];
  const pkgSpecArr = pkgSpec.split('/');
  return {
    name: pkgSpecArr[0],
    version: pkgSpecArr[1],
  };
}

function isEmptyDir(path: string) {
  const files = fs.readdirSync(path);
  return files.length === 0 || (files.length === 1 && files[0] === '.git');
}

export async function main() {
  console.log('');
  logger.greet('â—†  Create Rsbuild Project');

  const cwd = process.cwd();
  const pkgInfo = pkgFromUserAgent(process.env.npm_config_user_agent);
  const pkgManager = pkgInfo ? pkgInfo.name : 'npm';
  const packageRoot = path.resolve(__dirname, '..');
  const packageJsonPath = path.join(packageRoot, 'package.json');
  const { version } = require(packageJsonPath);

  let targetDir = checkCancel<string>(
    await text({
      message: 'Input target folder',
      placeholder: 'my-project',
      validate(value) {
        if (value.length === 0) {
          return 'Target folder is required';
        }
      },
    }),
  );

  targetDir = formatTargetDir(targetDir);
  const distFolder = path.join(cwd, targetDir);

  if (fs.existsSync(distFolder) && !isEmptyDir(distFolder)) {
    const option = checkCancel<string>(
      await select({
        message: `"${targetDir}" is not empty, please choose:`,
        options: [
          { value: 'yes', label: 'Continue and override files' },
          { value: 'no', label: 'Cancel operation' },
        ],
      }),
    );

    if (option === 'no') {
      cancelAndExit();
    }
  }

  const framework = checkCancel<string>(
    await select({
      message: 'Select framework',
      options: [
        { value: 'react', label: 'React' },
        { value: 'vue3', label: 'Vue 3' },
        { value: 'vue2', label: 'Vue 2' },
        { value: 'lit', label: 'Lit' },
        { value: 'preact', label: 'Preact' },
        { value: 'svelte', label: 'Svelte' },
        { value: 'solid', label: 'Solid' },
        { value: 'vanilla', label: 'Vanilla' },
      ],
    }),
  );

  const language = checkCancel<string>(
    await select({
      message: 'Select language',
      options: [
        { value: 'ts', label: 'TypeScript' },
        { value: 'js', label: 'JavaScript' },
      ],
    }),
  );

  const tools = checkCancel<string[]>(
    await multiselect({
      message: 'Select additional tools (use arrow keys / space bar)',
      options: [
        { value: 'prettier', label: 'Add Prettier for code formatting' },
      ],
    }),
  );

  const srcFolder = path.join(packageRoot, `template-${framework}-${language}`);
  const commonFolder = path.join(packageRoot, 'template-common');

  copyFolder(commonFolder, distFolder, version);
  copyFolder(srcFolder, distFolder, version, path.basename(targetDir));

  for (const tool of tools) {
    const toolFolder = path.join(packageRoot, `template-${tool}`);
    copyFolder(toolFolder, distFolder, version);
  }

  const nextSteps = [
    `cd ${targetDir}`,
    `${pkgManager} i`,
    `${pkgManager} run dev`,
  ];

  note(nextSteps.join('\n'), 'Next steps');

  outro('Done.');
}

function sortObjectKeys(obj: Record<string, unknown>) {
  const sortedKeys = Object.keys(obj).sort();

  const sortedObj: Record<string, unknown> = {};
  for (const key of sortedKeys) {
    sortedObj[key] = obj[key];
  }

  return sortedObj;
}

function mergePackageJson(targetPackage: string, extraPackage: string) {
  if (!fs.existsSync(targetPackage)) {
    return;
  }

  const content: Record<string, unknown> = deepmerge(
    JSON.parse(fs.readFileSync(targetPackage, 'utf-8')),
    JSON.parse(fs.readFileSync(extraPackage, 'utf-8')),
  );

  for (const key of ['scripts', 'dependencies', 'devDependencies']) {
    if (!(key in content)) {
      continue;
    }
    content[key] = sortObjectKeys(content[key] as Record<string, unknown>);
  }

  fs.writeFileSync(targetPackage, `${JSON.stringify(content, null, 2)}\n`);
}

function copyFolder(src: string, dist: string, version: string, name?: string) {
  const renameFiles: Record<string, string> = {
    gitignore: '.gitignore',
  };

  // Skip local files
  const skipFiles = ['node_modules', 'dist'];

  fs.mkdirSync(dist, { recursive: true });

  for (const file of fs.readdirSync(src)) {
    if (skipFiles.includes(file)) {
      continue;
    }

    const srcFile = path.resolve(src, file);
    const distFile = renameFiles[file]
      ? path.resolve(dist, renameFiles[file])
      : path.resolve(dist, file);
    const stat = fs.statSync(srcFile);

    if (stat.isDirectory()) {
      copyFolder(srcFile, distFile, version);
    } else if (file === 'extra-package.json') {
      const targetPackage = path.resolve(dist, 'package.json');
      mergePackageJson(targetPackage, srcFile);
    } else {
      fs.copyFileSync(srcFile, distFile);

      if (file === 'package.json') {
        updatePackageJson(distFile, version, name);
      }
    }
  }
}

const updatePackageJson = (
  pkgJsonPath: string,
  version: string,
  name?: string,
) => {
  let content = fs.readFileSync(pkgJsonPath, 'utf-8');
  content = content.replace(/workspace:\*/g, `^${version}`);
  const pkg = JSON.parse(content);

  if (name && name !== '.') {
    pkg.name = name;
  }

  fs.writeFileSync(pkgJsonPath, JSON.stringify(pkg, null, 2));
};

main();
