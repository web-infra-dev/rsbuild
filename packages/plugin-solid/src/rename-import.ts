import { createVirtualModule } from '@rsbuild/shared';
import { readFileSync } from 'fs';
import type { PluginObj } from '@babel/core';

export default function renameImport(): PluginObj {
  return {
    visitor: {
      ImportDeclaration(path) {
        if (path.node.source.value === 'solid-refresh') {
          const runtimeFilePath = require.resolve(
            'solid-refresh/dist/solid-refresh.mjs',
          );
          const runtimeCode = readFileSync(runtimeFilePath, 'utf-8');
          path.node.source.value = createVirtualModule(runtimeCode);
        }
      },
    },
  };
}
