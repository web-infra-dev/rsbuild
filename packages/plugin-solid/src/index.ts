import { PLUGIN_BABEL_NAME, type RsbuildPlugin } from '@rsbuild/core';
import type { SolidPresetOptions } from './types';
import { modifyBabelLoaderOptions } from '@rsbuild/plugin-babel';

export type PluginSolidPresetOptions = {
  solidPresetOptions?: SolidPresetOptions;
};

export const PLUGIN_SOLID_NAME = 'rsbuild:solid';

export function pluginSolid(
  options: PluginSolidPresetOptions = {},
): RsbuildPlugin {
  return {
    name: PLUGIN_SOLID_NAME,

    pre: [PLUGIN_BABEL_NAME],

    setup(api) {
      api.modifyBundlerChain(async (chain, { CHAIN_ID, isProd }) => {
        const rsbuildConfig = api.getNormalizedConfig();

        modifyBabelLoaderOptions({
          chain,
          CHAIN_ID,
          modifier: (babelOptions) => {
            babelOptions.presets ??= [];
            babelOptions.presets.push([
              require.resolve('babel-preset-solid'),
              options.solidPresetOptions || {},
            ]);

            if (!isProd && rsbuildConfig.dev.hmr) {
              babelOptions.plugins ??= [];
              babelOptions.plugins.push([
                require.resolve('solid-refresh/babel'),
              ]);

              chain.resolve.alias.merge({
                'solid-refresh': require.resolve(
                  'solid-refresh/dist/solid-refresh.mjs',
                ),
              });
            }

            return babelOptions;
          },
        });
      });
    },
  };
}
