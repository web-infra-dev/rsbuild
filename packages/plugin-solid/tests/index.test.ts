import { createStubRsbuild } from '@rsbuild/test-helper';
import { pluginBabel } from '@rsbuild/plugin-babel';
import { pluginSolid } from '../src';

describe('plugin-solid', () => {
  it('should apply solid preset correctly in rspack mode', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [pluginSolid(), pluginBabel()],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should apply solid preset correctly', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [pluginSolid(), pluginBabel()],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });

  it('should allow to configure solid preset options', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [
        pluginSolid({
          solidPresetOptions: {
            generate: 'ssr',
            hydratable: true,
          },
        }),
        pluginBabel(),
      ],
    });
    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
