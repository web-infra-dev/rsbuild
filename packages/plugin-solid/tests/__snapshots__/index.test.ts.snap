// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-solid > should allow to configure solid preset options 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-babel/compiled/babel-loader/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/babel.cjs",
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/babel-preset-solid/index.js",
                  {
                    "generate": "ssr",
                    "hydratable": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": {
    "alias": {
      "solid-refresh": "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/solid-refresh.mjs",
    },
  },
}
`;

exports[`plugin-solid > should apply solid preset correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-babel/compiled/babel-loader/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/babel.cjs",
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/babel-preset-solid/index.js",
                  {},
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": {
    "alias": {
      "solid-refresh": "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/solid-refresh.mjs",
    },
  },
}
`;

exports[`plugin-solid > should apply solid preset correctly in rspack mode 1`] = `
{
  "module": {
    "rules": [
      {
        "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-babel/compiled/babel-loader/index.js",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/babel.cjs",
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/babel-preset-solid/index.js",
                  {},
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": {
    "alias": {
      "solid-refresh": "<ROOT>/node_modules/<PNPM_INNER>/solid-refresh/dist/solid-refresh.mjs",
    },
  },
}
`;
